airflow/providers/common/sql/hooks/sql.py

```diff
@@ -218,6 +218,26 @@ def get_pandas_df(self, sql, parameters: Iterable | Mapping[str, Any] | None = N
         with closing(self.get_conn()) as conn:
             return psql.read_sql(sql, con=conn, params=parameters, **kwargs)
 
+    def get_polars_df(self, sql, **kwargs):
+        """
+        Executes the sql and returns a polars dataframe.
+
+        :param sql: the sql statement to be executed (str) or a list of
+            sql statements to execute
+        :param parameters: The parameters to render the SQL query with.
+        :param kwargs: (optional) passed into polars.read_database method
+        """
+        try:
+            import polars as pl
+        except ImportError:
+            raise Exception(
+                "polars library not installed, run: pip install "
+                "'apache-airflow-providers-common-sql[polars]'."
+            )
+
+        with closing(self.get_conn()) as conn:
+            return pl.read_database(sql, connection=conn, **kwargs)
```

Comment:
I'm not sure how for `polaris`, but for `pandas` the same method is broken. Let me explain:

1. Pandas expect to get SQLAlchemy connection rather than DBAPI with only one  exception for SQLite
2. `get_sqlalchemy_engine` is also broken in some cases because `get_url` is broken, see discussion in dev list: https://lists.apache.org/thread/8rhmz3qh30hvkondct4sfmgk4vd07mn5

Reply:
Maybe it is not a case for `polaris` I'm not familiar with this lib

Reply:
In the issue it was suggested to have `get_df` (generic function) that accept as parameter if it should be pandas/polars - why did you choose not to have it eventually?

Reply:
I was suggest that we should not merge something in single implementation especially if no one know how `pandas` and `polaris` interface compatible. I just worried that if we merge it now, than it could turned into the something like [BackfillJobRunner._backfill_job_runner.py](https://github.com/apache/airflow/blob/7a5b6a39c948bc51d754898b0478ad319dfab43f/airflow/jobs/backfill_job_runner.py#L432) which literally have 170 different statements.

And in additional pandas is broken for at least half implementations:  https://github.com/apache/airflow/pull/34679#discussion_r1341200518, so my vision that we should fix it first before we could create some generic method

Reply:
@Taragolis [the `polars` doc on `read_database()`](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.read_database.html) states the following:
```
This function supports a wide range of native database drivers (ranging from local databases such as SQLite to large cloud databases such as Snowflake), as well as generic libraries such as ADBC, SQLAlchemy and various flavours of ODBC. 
```
... and the `connection` object eventually gets passed in the constructor to the `polars` `ConnectionExecutor` [here](https://github.com/pola-rs/polars/blob/fda3073826eb99dcfcb8f35bebd3eae74310d670/py-polars/polars/io/database.py#L78). Does this answer your question?

Reply:
> In the issue it was suggested to have `get_df` (generic function) that accept as parameter if it should be pandas/polars - why did you choose not to have it eventually?

@eladkal you are right ([link to suggestion in issue](https://github.com/apache/airflow/issues/33911#issuecomment-1699820326)), this was suggested. I made this decision because the APIs of `polars` and `pandas` are actually quite syntactically different for the actual querying, so the `get_df` generic function would have just been a big switch statement, which seemed to me to add no value. Furthermore, I'm not sure why we'd want to add `get_df` without also removing (or at least removing exposure of) the underlying methods `get_polars_df` and `get_pandas_df`; having `get_df`, `get_polars_df`, and `get_pandas`. Doing this removal, however, would be a breaking change for a lot of users' code. It seemed to me like it may fall better under the scope of a refactoring ticket (e.g. [as you say, there's legacy code on the import of pandas for backwards compatibility](https://github.com/apache/airflow/pull/34679#discussion_r1341134939)).

Reply:
We don't have to remove anything. We can deprecate and raise deprecation warning.
For such deprecation we probably will give very long time before actually removing it

Reply:
It still seems to me to be over-engineering, at least for now @eladkal .